docs/qml/qaoa.md

```diff
@@ -34,57 +36,111 @@ The goal of the MaxCut algorithm is to maximize the following cost function:
+)
+
+```
+
 ## The QAOA quantum circuit
 
-This problem can be solved by using a parametrized quantum circuit with the QAOA algorithm. This
-requires a circuit with two main components:
+The Max-Cut problem can be solved by using the QAOA algorithm. QAOA belongs to the class of Variational Quantum Algorithms (VQAs), which means that its quantum circuit contains a certain number of parametrized quantum gates that need to be optimized with a classical optimizer.
+The QAOA circuit is composed of two operators:
 
-* the **cost component**: a circuit generated by a diagonal Hamiltonian which
-  encodes the cost function described above into a quantum circuit.
-* the **mixing component**: a simple set of single-qubit rotations with adjustable
+* The **cost operator $U_c$**: a circuit generated by the cost Hamiltonian which
+encodes the cost function described above into a quantum circuit. The solution to the optimization problem is encoded in the ground state of the cost Hamiltonian $H_c$.
+The cost operator  is simply the evolution of the cost Hamiltonian parametrized by a variational parameter $\gamma$ so that $U_c = e^{i\gamma H_c}.$
+* The **mixing operator $U_b$**: a simple set of single-qubit rotations with adjustable
   angles which are tuned during the classical optimization loop to minimize the cost
 
-Below, the QAOA quantum circuit with the cost and mixing components is defined using
-`qadence` operations. The cost component of each layer of the circuit is decomposed
+
+The QAOA quantum circuit consists of a number of layers, each layer containing a cost and a mixing operator.
+Below, the QAOA quantum circuit is defined using
+`qadence` operations.
+First, a layer of Hadamard gates is applied to all qubits to prepare the initial state $|+\rangle ^{\otimes n}$.
+The cost operator of each layer can be built "manually", implementing the $e^{iZZ\gamma}$ terms with CNOTs and a $\rm{RZ}(2\gamma)$ rotation, or it can also be automatically decomposed
 into digital single and two-qubits operations via the `.digital_decomposition()` method.
 The decomposition is exact since the Hamiltonian generator is diagonal.
 
 ```python exec="on" source="material-block" html="1" session="qaoa"
-from qadence import Zero, I, HamEvo, tag, kron, chain, QuantumCircuit, RX, Z
-
-# generators associated with the edges of the given graph
-zz_ops = [kron(Z(edge[0]), Z(edge[1])) for edge in graph.edges()]
+from qadence import tag, kron, chain, RX, RZ, Z, H, CNOT, Zero, I
+from qadence import HamEvo, QuantumCircuit, Parameter
 
 n_qubits = graph.number_of_nodes()
-n_layers = 2
+n_edges = graph.number_of_edges()
+n_layers = 6
+
 
+# Generate the cost Hamiltonian
+zz_ops = [kron(Z(edge[0]), Z(edge[1])) for edge in graph.edges()]
 cost_ham = Zero()
 for op in zz_ops:
-    cost_ham += 0.5 * op
-cost_ham = 0.5 * kron(I(i) for i in range(n_qubits)) - cost_ham
+    cost_ham += op
+cost_ham = 0.5 * (n_edges * kron(I(i) for i in range(n_qubits)) - cost_ham)
```

Comment:
It can indeed be written in terms of of the number operator $N=\frac{1}{2}(1-Z)$, but then the formulation is a bit less straightforward and it is not the usual way to formulate MaxCut in this setting. I think that with the number operator the Hamiltonian would be:

$$ H = \sum_{\langle i,j \rangle} N_i + N_j - 2N_i N_j$$

where $\langle i,j \rangle$ are the graph edges. Note that the only difference between the two formulations is that the number operator acts on the zero state as $N|0\rangle = 0$ and the Z operator as $Z|0\rangle = -1$, while they both act equally on the $|1\rangle$. I think it might be better to keep it with the Z operators since it's the usual way to formulate it.